automata/fa/dfa.py

```diff
@@ -598,6 +598,79 @@ def isfinite(self):
         except nx.exception.NetworkXNoCycle:
             return True
 
+    @classmethod
+    def contains_subsequence(cls, subsequence, input_symbols):
+        """
+        Creates a DFA which accepts all words which contain a specific subsequence of symbols
+        """
+        transitions = {0: {symbol: 0 for symbol in input_symbols}}
+
+        for prev_state, symbol in enumerate(subsequence):
+            next_state = prev_state + 1
+            transitions[next_state] = {symbol: next_state for symbol in input_symbols}
+            transitions[prev_state][symbol] = next_state
+
+        return cls(
+            states=set(transitions.keys()),
+            input_symbols=input_symbols,
+            transitions=transitions,
+            initial_state=0,
+            final_states={len(subsequence)},
+        )
+
+    @classmethod
+    def of_length(cls, min_length, max_length, input_symbols):
+        """
+        Creates a DFA which accepts all words whose length is between `min_length` and `max_length`, inclusive.
+        To allow infinitely long words the value `float('inf')` can be passed in for `max_length`.
+        """
+        transitions = {}
+        length_range = range(min_length) if max_length == float('inf') else range(max_length+1)
+        for prev_state in length_range:
+            next_state = prev_state + 1
+            transitions[prev_state] = {symbol: next_state for symbol in input_symbols}
+        last_state = len(transitions)
+        transitions[last_state] = {symbol: last_state for symbol in input_symbols}
+        final_states = {last_state} if max_length == float('inf') else set(range(min_length, max_length+1))
+        return cls(
+            states=set(transitions.keys()),
+            input_symbols=input_symbols,
+            transitions=transitions,
+            initial_state=0,
+            final_states=final_states,
+        )
+
+    @classmethod
+    def universal_language(cls, input_symbols):
+        return cls.contains_subsequence('', input_symbols)
+
+    @classmethod
+    def empty_language(cls, input_symbols):
+        return ~cls.universal_language(input_symbols)
+
+    @classmethod
+    def nth_from_end(cls, symbol, n, input_symbols):
+        """
+        Creates a DFA which accepts all words whose `n`-th character from the end is `symbol`,
+        where `n` is a positive integer.
+        """
+        # TODO: special case for len(input_symbols) == 1?
+        if n < 1:
+            raise ValueError("Integer must be positive")
+        if symbol not in input_symbols:
+            raise ValueError("Desired symbol is not in the set of input symbols")
+        state_count = 2**n
```

Comment:
Doesn't this only work if there are two input symbols?

Reply:
You have the symbol `symbol` which is the one you are looking for to be the start of your $n$ character suffix.
All other symbols belong to the same equivalence class of *not what you're looking for*. Therefore the construction looks something like this:
![image](https://user-images.githubusercontent.com/7704746/198084117-cca8bbe5-528b-4f28-8594-4b0d5af1c819.png)

Here we are looking for `x` and can treat all other characters as `?`

Reply:
Ahhhh I see now, thanks for the clarification 😃

Reply:
We might want to leave a comment in the code regarding this as it is not immediately obvious.

Reply:
@Tagl Agreed about adding a code comment (my future self will probably find it helpful).

Reply:
Comment added